Formatted output for XML patchsets

nrwahl2 opened the pull request:
This adds support for formatted output (`pcmk__output_t` message functions) for showing XML patchsets. The new message functions replace `pcmk__xml_log_patchset()`.

nrwahl2 commented on lib/common/output_xml.c:
```diff
@@ -467,7 +467,7 @@ pcmk__output_xml_add_node_copy(pcmk__output_t *out, xmlNodePtr node) {
     CRM_CHECK(parent != NULL, return);
 
-    add_node_copy(parent, copy_xml(node));
+    add_node_copy(parent, node);
```
All regression tests passed with this change. I still wonder if I used `copy_xml()` to fix some arcane XML doc issue when filing https://github.com/ClusterLabs/pacemaker/commit/b64368eb96a059ae62c414f5e373a78487829428, or if the `copy_xml()` is just a leftover that I forgot to remove. Obviously this commit assumes the latter.

nrwahl2 commented on lib/common/patchset_display.c:
```diff
@@ -132,27 +147,104 @@ xml_show_patchset_v1(pcmk__output_t *out, const xmlNode *patchset)
+static int
+xml_patchset_v1_default(pcmk__output_t *out, va_list args)
+{
+    xmlNode *patchset = va_arg(args, xmlNode *);
```
This can be const for default and log. It just has to be non-const for XML due to doc stuff.

kgaillot commented on lib/common/output.c:
```diff
@@ -168,6 +170,12 @@ pcmk__register_messages(pcmk__output_t *out, const pcmk__message_entry_t *table)
+pcmk__register_common_messages(pcmk__output_t *out)
+{
+    pcmk__register_patchset_messages(out);
+}
```
I wonder whether it would be worthwhile to just do this in `pcmk__output_new()`

nrwahl2 replied:
That sounds reasonable

kgaillot commented on lib/common/xml_display.c:
```diff
@@ -139,29 +146,36 @@ show_xml_element(pcmk__output_t *out, GString *buffer, const char *prefix,
+            int temp_rc = show_xml_node(out, buffer, prefix, child, depth + 1,
+                                         options
+                                         |pcmk__xml_fmt_open
+                                         |pcmk__xml_fmt_close);
```
indenting

nrwahl2 replied:
Whoops, had it as `child_rc` at first.

kgaillot commented on lib/cib/cib_utils.c:
```diff
@@ -328,9 +328,10 @@ cib_perform_op(const char *op, int call_options, cib_op_t * fn, gboolean is_quer
-            pcmk__xml_show_changes(out, scratch);
-            out->finish(out, CRM_EX_OK, true, NULL);
+            rc = pcmk__xml_show_changes(out, scratch);
+            out->finish(out, pcmk_rc2exitc(rc), true, NULL);
```
pcmk_rc_no_output -> CRM_EX_IOERR, we probably want to remap that to ok (and anywhere else we'll be using rc2exitc)

nrwahl2 replied:
That mapping is questionable since `pcmk_rc_no_output` is a perfectly normal consequence of `out->is_quiet() == true`. I'm sure there are error conditions that can cause `pcmk_rc_no_output` too.

clumens replied:
No, I don't remember why we would have done that.  Change it and see what breaks, I guess?

kgaillot replied:
It was done when the new code was first added, and neither the commit message nor the PR (#2008) comment on it, so I'm guessing it wasn't chosen for any particular reason. Let's just map it to CRM_EX_OK in pcmk_rc2exitc() and run the regression tests.

nrwahl2 replied:
`cts-regression` passes with the mapping to `CRM_EX_OK`

kgaillot commented on lib/common/patchset_display.c:
```diff
@@ -132,27 +147,104 @@ xml_show_patchset_v1(pcmk__output_t *out, const xmlNode *patchset)
+{
+    xmlNode *patchset = va_arg(args, xmlNode *);
+
+    pcmk__output_xml_add_node_copy(out, patchset);
```
The benefit of a CDATA block is the API schema can just use `<text/>` for it rather than need to know the separate schema for what's being added.

nrwahl2 replied:
We already have something like an `any.rng`, and I was considering using that as an alternative to CDATA... somewhere! I wrote it, I thought as part of the rule schema update, but I can't find it in the source tree.

nrwahl2 replied:
The mystery will remain unsolved. No real reason to make one now (rather than using CDATA) if we don't already have one.

nrwahl2 replied:
I was using the `any-element` schema in this `diff-2.29.rng` schema, and I definitely linked to the gist within some other PR a couple months ago (don't know which one). But those schemas themselves haven't been part of any PR yet.

kgaillot reviewed (commented):
nice start

nrwahl2 commented on lib/common/patchset_display.c:
```diff
@@ -0,0 +1,317 @@
+/*
+ * Copyright 2023 the Pacemaker project contributors
```
Ah, didn't realize we did it that way. In that case I've made the same mistake before (including but possibly not limited to `xml_display.c`).

nrwahl2 commented:
I thought I was gonna finish up a draft tonight, but it looks like this isn't gonna work unless we can make a copy of a `const xmlNode *` and/or eliminate one or more `getDocPtr()` dependencies.

`libXml2` really screwed us here by making almost all of their functions non-const.

As of right now all of our XML copying functions and xml2text functions require non-const arguments. So the patchset message functions have to take a non-const `xmlNode *` argument. This means all the callers have to pass in a non-const. At least one caller (`xml_log_changes()`) is public API, albeit deprecated, that accepts a `const xmlNode *` argument. As mentioned above, making a copy requires that it be non-const.

I'm sure there's a way around this but it's a less straightforward than expected.

nrwahl2 commented:
I thought I was gonna finish up a draft tonight, but it looks like this isn't gonna work unless we can make a copy of a `const xmlNode *` and/or eliminate one or more `getDocPtr()` dependencies.

`libXml2` really screwed us here by making almost all of their functions non-const.

As of right now all of our XML copying functions and xml2text functions require non-const arguments. So the patchset message functions have to take a non-const `xmlNode *` argument. This means all the callers have to pass in a non-const. At least one caller (`xml_log_changes()`) is public API, albeit deprecated, that accepts a `const xmlNode *` argument. As mentioned above, making a copy requires that it be non-const.

I'm sure there's a way around this but it's a less straightforward than expected.

Edit: Looks like that's the only caller with `const`, so I can probably just duplicate some code into the deprecated function body to make it behave the old way and keep the const there.

kgaillot commented:
For cases where the XML document has already been created, getDocPtr() boils down to xml->doc. We can just use that anywhere we're confident the document already exists. Anything created by create_xml_node() (which I believe is everything we do, with the possible exception of some XML output formatter stuff) definitely has a document already. We can assert doc is not NULL where needed.

kgaillot commented:
libxml2 is more of a pain, but we have cast the const away in some of those cases already. It's not ideal since we can't guarantee they aren't actually changing anything, but if it's a real simple function and the current implementation doesn't change anything, it can be justified.

nrwahl2 commented:
I hesitate to do that for `xml_log_patchset()` since it accepts an arbitrary `xmlNode *` and is public, so we can't guarantee that it has a document. It's HIGHLY unlikely that external code is calling this, especially with a doc-less node, but I'd hate to start asserting if so.

That option is on the table if duplication doesn't solve it; I'm thinking I can get it done without being too ugly. Will have to see how the pieces fit together.

nrwahl2 commented:
Pushed some very-WIP stuff to the wrong branch here instead of my test branch... oh well, just ignore it for now.

nrwahl2 commented:
Replaced commits to address review, change the approach somewhat in replacing `pcmk__xml_log_patchset()`, and finish the work. The first several commits have few or no changes, but it'd be good to skim over them for sanity.

nrwahl2 commented:
Unit tests were failing. I updated "New pcmk__register_patchset_messages()" commit to modify the unit tests, so that we don't assume that a newly initialized `pcmk__output_t` object has 0 messages registered.

This is questionable... it does prevent duplication at the caller (see https://github.com/ClusterLabs/pacemaker/pull/3025#discussion_r1102144317 -- the original approach was to define a `pcmk__register_common_messages()` and call it before registering other messages). But it means the unit test code effectively has to register messages (via `pcmk__output_new()`) before it can test registering messages.

kgaillot commented on daemons/based/based_callbacks.c:
```diff
@@ -1430,7 +1430,9 @@ cib_process_command(xmlNode * request, xmlNode ** reply, xmlNode ** cib_diff, gb
 
-    pcmk__xml_log_patchset(LOG_TRACE, *cib_diff);
+    pcmk__output_set_log_level(logger_out, LOG_TRACE);
+    logger_out->message(logger_out, "xml-patchset", *cib_diff);
```
Where we're using this approach, do we need to reset the log level to whatever it originally was, or do all the uses already explicitly set the desired level?

nrwahl2 replied:
All three of the uses explicitly set the desired level. It seems much clearer and less error-prone that way, unless we need something more complex later where we set the log level for a whole block and something called within the block also needs to set the log level.

kgaillot commented on lib/cib/cib_file.c:
```diff
@@ -869,29 +869,39 @@ cib_file_perform_op_delegate(cib_t * cib, const char *op, const char *host, cons
+    }
 
-    } else if(output != in_mem_cib) {
+done:
```
Are you going to need this label later, or is it a stray?

nrwahl2 replied:
I'm using it now, in the `else if (query == FALSE)` block. That way I can break out, free variables, and return if there's an output initialization problem.

kgaillot commented on tools/crm_diff.c:
```diff
@@ -209,13 +209,20 @@ generate_patch(xmlNode *object_1, xmlNode *object_2, const char *xml_file_2,
         XML_ATTR_NUMUPDATES,
     };
 
+    rc = pcmk__log_output_new(&out);
```
Is this going to conflict when you add formatted output to crm_diff? The code here will always want a logger, whereas the overall output object could have any format

nrwahl2 replied:
It probably will. I was thinking the non-logger `out` object could be local to `main()`, but a handful of helper functions use it, so it'll be better to make it global.

I could change the name of this variable later when we get to that point, but since I have a couple more changes to make prior to merge, I might as well do it now.

kgaillot reviewed (commented):
progress

nrwahl2 commented on lib/common/patchset.c:
```diff
@@ -18,9 +18,7 @@
 
-#include <libxml/parser.h>
 #include <libxml/tree.h>
-#include <libxml/xmlIO.h>  /* xmlAllocOutputBuffer */
```
True. Either way, we should be good on these two -- it looked like nothing from them was used in this file.

nrwahl2 commented on lib/common/output.c:
```diff
@@ -84,6 +86,9 @@ pcmk__output_new(pcmk__output_t **out, const char *fmt_name, const char *filenam
 
+    // Register libcrmcommon messages (currently they exist only for patchset)
+    pcmk__register_patchset_messages(*out);
+
```
That sounds reasonable

nrwahl2 commented on lib/fencing/st_output.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2019-2022 the Pacemaker project contributors
+ * Copyright 2019-2023 the Pacemaker project contributors
```
Yep, thought I got rid of all those

nrwahl2 commented on include/crm/common/xml_internal.h:
```diff
@@ -164,8 +164,8 @@ enum pcmk__xml_fmt_options {
-void pcmk__xml_show(pcmk__output_t *out, const char *prefix,
-                    const xmlNode *data, int depth, uint32_t options);
+int pcmk__xml_show(pcmk__output_t *out, const char *prefix, const xmlNode *data,
+                   int depth, uint32_t options);
```
Good catch, I either got my wires crossed editing the message or forgot to edit it completely when the approach changed.

nrwahl2 commented on lib/common/patchset_display.c:
```diff
@@ -106,21 +123,21 @@ xml_show_patchset_v1_recursive(pcmk__output_t *out, const char *prefix,
-    }
+    int rc = xml_show_patchset_header(out, patchset);
 
     xml_show_patchset_header(out, patchset);
```
Whoops, definitely. I separated this commit manually after I had already done the "Use message functions" commit, when I saw how big it had become.

nrwahl2 commented on lib/common/patchset_display.c:
```diff
@@ -301,19 +348,27 @@ pcmk__xml_log_patchset(uint8_t log_level, const xmlNode *patchset)
             break;
         default:
             crm_err("Unknown patch format: %d", format);
+            rc = pcmk_rc_unknown_format;
```
By the way, this is a misuse of the return code the way it's currently mapped to a string in `results.c`: `"Unknown output format"`. It seems reasonable to use it more broadly, so maybe we could change that to simply say `"Unknown format"`. It's already being used in ways other than unknown output format throughout libcrmcommon.

kgaillot replied:
I'd rather keep unknown_format for its original purpose (--output-as set to something unrecognized), and add a new code or codes for all the current uses that don't fit that (looks like we were just being lazy). However it's not important enough to worry about at this point, I'll just create a task anyone can pick up later.

nrwahl2 commented:
Updated per review.
* New `pcmk__bare_output_new()` to keep unit tests simple
* Removed stray copyright update from `st_output.c`
* Fixed commit message `s/any call/all calls/`
* Removed duplicate `xml_show_patchset_header()` call
* Renamed `crm_diff`'s logger to `logger_out` to avoid future conflicts

I kept `pcmk_rc_unknown_format` for now as discussed

nrwahl2 commented:
Added `#include "../../crmcommon_private.h"` to the affected unit test files.
